### PR TITLE
[Docs] Translate the migration/hook.md to English

### PR DESCRIPTION
**Pull request by SheffieldCao**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Update docs: English version of migration/hook.md.

## Modification

Add English version doc migration/hook.md.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/en/migration/hook.md [hunk 1]
```diff
@@ -1,3 +1,319 @@
 # Migrate Hook from MMCV to MMEngine
 
-Coming soon. Please refer to [chinese documentation](https://mmengine.readthedocs.io/zh_CN/latest/migration/hook.html).
+## Introduction
+
+Due to the development of our architecture design and increasing user demands, existing hook mount points in MMCV are insufficient for the needs. Hence, we redesigned the mount points in MMEngine, and the functions of hooks are adjusted accordingly. Before your migration, it will help a lot to read the tutorial [Hook Design](../design/hook.md).
```
Comment (HAOCHENYE) [inline docs/en/migration/hook.md:5]:
```suggestion
Due to the upgrade of our architecture design and increasing user demands, existing hook mount points in MMCV can no logger meet the requirements. Hence, we redesigned the mount points in MMEngine, and the functions of hooks are adjusted accordingly. It will help a lot to read the tutorial [Hook Design](../design/hook.md) before your migration.
```

File: docs/en/migration/hook.md [hunk 2]
```diff
@@ -1,3 +1,319 @@
...
+
+This tutorial compares the difference in function, mount point, usage and implementation between the hooks from [MMCV v1.6.0](https://github.com/open-mmlab/mmcv/tree/v1.6.0) and [MMEngine v0.5.0](https://github.com/open-mmlab/mmengine/tree/v0.5.0).
```
Comment (HAOCHENYE) [inline docs/en/migration/hook.md:7]:
```suggestion
This tutorial compares the difference in function, mount point, usage and implementation between [MMCV v1.6.0](https://github.com/open-mmlab/mmcv/tree/v1.6.0) and [MMEngine v0.5.0](https://github.com/open-mmlab/mmengine/tree/v0.5.0).
```

File: docs/en/migration/hook.md [hunk 3]
```diff
@@ -1,3 +1,319 @@
...
+    <td>after each iteration</td>
+    <td>None</td>
+    <td>after_test_iter，with additional args: batch_idx、data_batch and outputs</td>
+  </tr>
+</tbody>
+</table>
+
+## Usage Comparison
+
+In MMCV, to register hooks to the Runner, you need to call the `register_training_hooks` method of the Runner to register hooks on the Runner, while in MMEngine, you can pass parameters to the Runner's initialization method for registration.
```
Comment (HAOCHENYE) [inline docs/en/migration/hook.md:228]:
```suggestion
In MMCV, you need to call the `Runner.register_training_hooks` to register the hooks, while in MMEngine, you can initialize the hooks by simply passing the parameters of hooks during building the runner.
```

File: docs/en/migration/hook.md [hunk 4]
```diff
@@ -1,3 +1,319 @@
...
+    xxx,
+)
+runner.train()
+```
+
+For more details of MMEngine hooks, please refer to [Usage of Hooks](../tutorials/hook.md).
+
+## Implementation Comparison
+
+Taking `CheckpointHook` as an example, compared with [CheckpointHook](https://github.com/open-mmlab/mmcv/blob/v1.6.0/mmcv/runner/hooks/checkpoint.py) in MMCV, [CheckpointHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/checkpoint_hook.py) of MMEngine needs to implement the `after_val_epoch` method, since new CheckpointHook supports saving the optimal weights, while in MMCV, the function is achieved by EvalHook.
```
Comment (HAOCHENYE) [inline docs/en/migration/hook.md:288]:
```suggestion
Taking `CheckpointHook` as an example, compared with [CheckpointHook](https://github.com/open-mmlab/mmcv/blob/v1.6.0/mmcv/runner/hooks/checkpoint.py) in MMCV, [CheckpointHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/checkpoint_hook.py) of MMEngine needs to implement the `after_val_epoch` method, since new `CheckpointHook` supports saving the optimal weights, while in MMCV, the function is achieved by EvalHook.
```

File: docs/en/migration/hook.md [hunk 5]
```diff
@@ -1,3 +1,319 @@
...
+    <td>CheckpointHook</td>
+    <td rowspan="2">New CheckpointHook can save model weights and optimal weights as well. While the model evaluation function of EvalHook can be achieved by ValLoop or TestLoop</td>
+  </tr>
+  <tr>
+    <td>Model evaluation and optimal weights saving</td>
+    <td>EvalHook</td>
+  </tr>
+  <tr>
+    <td>Log printing</td>
+    <td rowspan="3">LoggerHook and its subclasses can print logs, save logs and visualize functions</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:48]:
```suggestion
    <td rowspan="3">LoggerHook and its subclasses can print logs, save logs and visualize data</td>
```

File: docs/en/migration/hook.md [hunk 6]
```diff
@@ -1,3 +1,319 @@
...
+    <td>Adding runtime information</td>
+    <td>RuntimeInfoHook</td>
+  </tr>
+  <tr>
+    <td>Model weights exponential moving average (EMA)</td>
+    <td>EMAHook</td>
+    <td>EMAHook</td>
+  </tr>
+  <tr>
+    <td>Ensuring the shuffle function of distributed Sampler</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:65]:
```suggestion
    <td>Ensuring the shuffle functionality of the distributed Sampler takes effect</td>
```

File: docs/en/migration/hook.md [hunk 7]
```diff
@@ -1,3 +1,319 @@
...
+    <td rowspan="2">Use ParamSchdulerHook and subclasses of <a href="../tutorials/param_scheduler.html">_ParamScheduler</a> to complete the adjustment of optimizer hyperparameters</td>
+  </tr>
+  <tr>
+    <td>Momentum adjustment</td>
+    <td>MomentumUpdaterHook</td>
+  </tr>
+  <tr>
+    <td>Saving model weights at specified interval</td>
+    <td>CheckpointHook</td>
+    <td rowspan="2">New CheckpointHook can save model weights and optimal weights as well. While the model evaluation function of EvalHook can be achieved by ValLoop or TestLoop</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:40]:
```suggestion
    <td rowspan="2">The CheckpointHook is responsible for not only saving weights but also saving the optimal weights. Meanwhile, the model evaluation function of EvalHook is delegated to ValLoop or TestLoop.</td>
```

File: docs/en/migration/hook.md [hunk 8]
```diff
@@ -1,3 +1,319 @@
...
+    <td>IterTimerHook</td>
+    <td>IterTimerHook</td>
+  </tr>
+  <tr>
+    <td>Analyzing bottlenecks of training time</td>
+    <td>ProfilerHook</td>
+    <td>Not yet available</td>
+  </tr>
+  <tr>
+    <td>Provide registration method to the function of hooks</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:90]:
```suggestion
    <td>Provide the most concise function registration</td>
```

File: docs/en/migration/hook.md [hunk 9]
```diff
@@ -1,3 +1,319 @@
...
+  </tr>
+  <tr>
+    <td>after each epoch</td>
+    <td>after_val_epoch</td>
+    <td>after_val_epoch</td>
+  </tr>
+  <tr>
+    <td>before each iteration</td>
+    <td>before_val_iter</td>
+    <td>before_val_iter，with additional args: batch_idx and data_batch</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:185]:
```suggestion
    <td>before_val_iter, with additional args: batch_idx and data_batch</td>
```

File: docs/en/migration/hook.md [hunk 10]
```diff
@@ -1,3 +1,319 @@
...
+  </tr>
+  <tr>
+    <td>after each iteration</td>
+    <td>after_val_iter</td>
+    <td>after_val_iter，with additional args: batch_idx、data_batch and outputs</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:190]:
```suggestion
    <td>after_val_iter, with additional args: batch_idx、data_batch and outputs</td>
```

File: docs/en/migration/hook.md [hunk 11]
```diff
@@ -1,3 +1,319 @@
...
+  </tr>
+  <tr>
+    <td>after each epoch</td>
+    <td>None</td>
+    <td>after_test_epoch</td>
+  </tr>
+  <tr>
+    <td>before each iteration</td>
+    <td>None</td>
+    <td>before_test_iter，with additional args: batch_idx and data_batch</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:216]:
```suggestion
    <td>before_test_iter, with additional args: batch_idx and data_batch</td>
```

File: docs/en/migration/hook.md [hunk 12]
```diff
@@ -1,3 +1,319 @@
...
+  </tr>
+  <tr>
+    <td>after each iteration</td>
+    <td>None</td>
+    <td>after_test_iter，with additional args: batch_idx、data_batch and outputs</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:221]:
```suggestion
    <td>after_test_iter, with additional args: batch_idx、data_batch and outputs</td>
```

File: docs/en/migration/hook.md [hunk 13]
```diff
@@ -1,3 +1,319 @@
...
+  </tr>
+  <tr>
+    <td>after each epoch</td>
+    <td>after_train_epoch</td>
+    <td>after_train_epoch</td>
+  </tr>
+  <tr>
+    <td>before each iteration</td>
+    <td>before_train_iter</td>
+    <td>before_train_iter，with additional args: batch_idx and data_batch</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:154]:
```suggestion
    <td>before_train_iter, with additional args: batch_idx and data_batch</td>
```

File: docs/en/migration/hook.md [hunk 14]
```diff
@@ -1,3 +1,319 @@
...
+  </tr>
+  <tr>
+    <td>after each iteration</td>
+    <td>after_train_iter</td>
+    <td>after_train_iter，with additional args: batch_idx、data_batch, and outputs</td>
```
Comment (zhouzaida) [inline docs/en/migration/hook.md:159]:
```suggestion
    <td>after_train_iter, with additional args: batch_idx、data_batch, and outputs</td>
```

**Comment by SheffieldCao** [comment]
All the review suggestions are updated in the new commit. [a7c48dc](https://github.com/open-mmlab/mmengine/pull/1054/commits/a7c48dcfa7b511ef89a4bc2cd1f4d078acd8e546)

**Review by zhouzaida — APPROVED** [review]
